Documentation for the current state of the world

mccheah opened the pull request:
Added a Kubernetes section to the docs. Follows similarly to how the YARN section was laid out.

Partially closes #3 but there still needs to be developer workflow build docs.

mccheah committed:
Documentation for the current state of the world.

mccheah committed:
Adding navigation links from other pages

iyanuobidele commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+
+For example, if the registry host is `registry-host` and the registry is listening on port 5000:
+
+    cd $SPARK_HOME
```
cd $SPARK_HOME/dist ?

mccheah replied:
This is documentation under the assumption that Spark was unpacked from a tarball, such as when it is downloaded from the Spark website.

mccheah replied:
We should also record dev-workflow docs somewhere, these aren't included in this PR just yet.

iyanuobidele replied:
I totally agree.

mccheah commented:
There's a bunch of things in these docs that might not be ideal states of the world even for the MVP, mostly little things. We can clean up the docs as the implementation evolves, though.

mccheah commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+  </td>
+</tr>
+<tr>
+  <td><code>spark.kubernetes.driver.uploads.driverExtraClasspath</code></td>
```
We probably shouldn't have this, I don't know how common using this will be.

iyanuobidele replied:
I believe this is fine.

mccheah commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+  `spark.kubernetes.driver.uploads.jars` in the application's configuration, will be treated as jars that are located on
+  the *disk of the submitting machine*. These jars are uploaded to the driver docker container before executing the
+  application.
+* A main application resource path that does not have a scheme or that has the scheme `file://` is assumed to be on the
```
It would be nice to be able to specify a main application resource on the container's disk as well. The main trouble here is how to specify that: do we create a new custom file scheme, like `docker://` to denote that the file is in the docker image?

iyanuobidele replied:
for sake of usability, I believe we should support the `docker://` scheme, since `'no scheme' == file://`

mccheah replied:
Yep - it's not particularly great that the API is to expect a magical prefix but I don't see a better option.

foxish replied:
Docker isn't the only runtime that is supported (although it is the most common), so, we could opt for something neutral like `container://`, or `pod://`.

ash211 replied:
Good point @foxish -- I like `container://` better, since in my understanding a k8s pod can have multiple containers each with independent filesystems so `pod://` isn't precise enough

mccheah replied:
If we were to include `container://` urls, we could also use this kind of URL scheme for the uploaded jars, and remove `spark.kubernetes.driver.uploads.jars`. I kind of like having the partitioning into two settings in this case though since `spark.jars` has preconceived expectations in all of the cluster managers, and there is some dissonance in making Kubernetes handle `spark.jars` in a special way. However this then makes specifying the main resource jar inconsistent with specifying the other uploaded jars.

mccheah commented:
Dev workflow docs probably belong under the `resource-managers/kubernetes` directory. I will follow up there in a separate PR.

iyanuobidele commented:
LGTM, great job @mccheah

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+---
+
+Support for running on [Kubernetes](https://kubernetes.io/) is available in experimental status. The feature set is
+currently limited and not well-tested.
```
let's lay it on a little stronger and make a note about this not being recommended for running in production

foxish commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+      --deploy-mode cluster 
+      --class com.example.applications.PluggableApplication
+      --master k8s://https://192.168.99.100
+      --kubernetes-namespace spark.kubernetes.namespace=default 
```
Specifying the `spark.kubernetes.namespace` along with the `--kubernetes-namespace` flag seems a little redundant.

mccheah replied:
Yep this is a typo

foxish commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+lifted in the future include:
+* Applications can only use a fixed number of executors. Dynamic allocation is not supported.
+* Applications can only run in cluster mode.
+* The external shuffle service cannot be used.
```
Dynamic allocation being unsupported also implies this I think.

foxish commented:
Looks very good overall @mccheah! Thanks for taking the effort.

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+
+## Setting Up Docker Images
+
+In order to run Spark on Kubernetes, a Docker image must be built and available on the Docker registry. Spark
```
"and available in an accessible Docker registry" (thinking about on-prem, internet-isolated use cases)

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+The Spark master, specified either via passing the `--master` command line argument to `spark-submit` or by setting
+`spark.master` in the application's configuration, must be a URL with the format `k8s://<api_server_url`. Prefixing the
+master string with `k8s://` will cause the Spark application to launch on a Kubernetes cluster, where the API server is
+contacted  at the appropriate inner URL. The HTTP protocol must also be specified.
```
do you mean **may** also be specified, with a mind towards non-SSL API servers?

mccheah replied:
It must be specified as a full URL currently, this is not ideal though and we should update when we change the default to `https`

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+master string with `k8s://` will cause the Spark application to launch on a Kubernetes cluster, where the API server is
+contacted  at the appropriate inner URL. The HTTP protocol must also be specified.
+
+Note that applications can currently only be executed in cluster mode.
```
"in cluster mode, where the Spark driver and its executors are running in the cluster"

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+ 
+### Adding Other JARs
+ 
+Spark allows users to provide dependencies that live on the driver's docker image, or that are on the local disk of the
```
capitalize Docker

ash211 replied:
"that live on" -> "that are bundled into"

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+* Local jars provided by specifying the `--upload-jars` command line argument to `spark-submit`, or by setting
+  `spark.kubernetes.driver.uploads.jars` in the application's configuration, will be treated as jars that are located on
+  the *disk of the submitting machine*. These jars are uploaded to the driver docker container before executing the
+  application.
```
"and are placed on the driver's classpath"

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+      --deploy-mode cluster
+      --class com.example.applications.SampleApplication
+      --master k8s://https://192.168.99.100
+      --kubernetes-namespace spark.kubernetes.namespace=default 
```
the `spark.kubernetes.namespace=` part seems redundant -- is that a typo?  I think you mixed it up with `--conf spark.kubernetes.namespace=default` ?

mccheah replied:
Yep typo

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+      --conf spark.kubernetes.driver.docker.image=registry-host:5000/spark-driver:latest
+      --conf spark.kubernetes.executor.docker.image=registry-host:5000/spark-executor:latest
+      /home/exampleuser/exampleapplication/main.jar
+
```
new paragraph:

note that `--upload-jars` is equivalent to `--conf spark.kubernetes.driver.uploads.jars`

If we can bold those lines in both indentations that would be nice too

mccheah replied:
Tricky to do both preformatted text and bolding in the same line; we'd have to invoke HTML for that. I think this is fine as is really.

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+      --conf spark.kubernetes.driver.docker.image=registry-host:5000/spark-driver-custom:latest
+      --conf spark.kubernetes.executor.docker.image=registry-host:5000/spark-executor:latest
+      http://example.com:8080/applications/sparkpluggable/app.jar
+      
```
same callout paragraph afterwards:

`--jars` is the same as `--conf spark.jars` and bold that line

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+<tr><th>Property Name</th><th>Default</th><th>Meaning</th></tr>
+<tr>
+  <td><code>spark.kubernetes.namespace</code></td>
+  <td>(none)</td>
```
do the other docs use `(none)` for required docs too?  I'd be inclined to put `(required)` here instead

mccheah replied:
We should probably just default to `default` however so that won't matter much in the longer term.

foxish replied:
+1 We should default to the `default` namespace.

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+  <td>(none)</td>
+  <td>
+    The namespace that will be used for running the driver and executor pods. Must be specified. When using
+    <code>spark-submit</code> in cluster mode, this can also be passed to <code>spark-submit></code> via the
```
extra `>` after the second `spark-submit`

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+</tr>
+<tr>
+  <td><code>spark.kubernetes.driver.docker.image</code></td>
+  <td><code>spark-driver:2.2.0</code></td>
```
let's change this from `2.2.0` to something like `2.2.0-latest` for now in both image versions.  I want to make sure we never accidentally publish a release that we'd later try to replace with something else in the same name.  `-latest` at least makes it clear the image can change

mccheah replied:
Currently the code refers to `SPARK_VERSION` to get the version here... which may actually be `2.2.0-SNAPSHOT` in this case. Not sure how to keep these docs in sync with the specific version string, but `2.2.0-SNAPSHOT` seems like the actual correct default here.

mccheah replied:
I'm ok with 2.2.0 for now but would be happy to hear suggestions in general. I think this will strictly depend on what we end up doing for publishing Docker images, if we publish at all. If we don't publish at all though then `spark-driver:latest` and `spark-executor:latest` seems reasonable for defaults, but defaults would rarely be used in practice because everyone would tag their own images slightly differently (`spark-driver:latest`, `spark_driver:2.2.0`, `sparkDriver:2.2.0`, `spark-driver:2.2.0`, `spark-driver:2.2`, `spark-app:2.2.0`...), not to mention sometimes needing to specify the Docker registry host in the tag.

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+  <td><code>spark.kubernetes.submit.caCertFile</code></td>
+  <td>(none)</td>
+  <td>
+    CA Cert file for connecting to Kubernetes over HTTPs.
```
lowercase `cert`.  uppercase `HTTPS`

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+  <td><code>spark.kubernetes.submit.clientKeyFile</code></td>
+  <td>(none)</td>
+  <td>
+    Client key file for authenticating against the Kubernetes API server.
```
are these paths?  just to local files, no https/http supported here?  let's make that explicit

mccheah replied:
These are files living on the local disk of the submitting machine.

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+  <td><code>spark.kubernetes.driver.uploads.driverExtraClasspath</code></td>
+  <td>(none)</td>
+  <td>
+    Comma-separated list of jars to be sent to the driver only when submitting the application in cluster mode. 
```
is that bit about only when submitting the application in cluster mode relevant?  k8s only supports cluster mode now

mccheah replied:
In client mode when we do end up supporting it, this won't really apply there. I think this is fine even if redundant as we don't have to change this part of the docs down the road.

ash211 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,206 @@
+  </td>
+</tr>
+</table>
+
```
Put an example here where we try to simplify the spark-submit command as much possible.  Something like:

---------

For example, the first example above can be rewritten from:

```
   bin/spark-submit
      --deploy-mode cluster
      --class com.example.applications.SampleApplication
      --master k8s://https://192.168.99.100
      --kubernetes-namespace spark.kubernetes.namespace=default 
      --upload-jars /home/exampleuser/exampleapplication/dep1.jar,/home/exampleuser/exampleapplication/dep2.jar
      --conf spark.kubernetes.driver.docker.image=registry-host:5000/spark-driver:latest
      --conf spark.kubernetes.executor.docker.image=registry-host:5000/spark-executor:latest
      /home/exampleuser/exampleapplication/main.jar
```

to

```
   bin/spark-submit
      --class com.example.applications.SampleApplication
      --upload-jars /home/exampleuser/exampleapplication/dep1.jar,/home/exampleuser/exampleapplication/dep2.jar
      /home/exampleuser/exampleapplication/main.jar
```

with these contents in `spark-defaults.conf`:

```
spark.master k8s://https://192.168.99.100
spark.submit.deployMode cluster
spark.kubernetes.namespace default
spark.kubernetes.driver.docker.image registry-host:5000/spark-driver:latest
spark.kubernetes.executor.docker.image registry-host:5000/spark-executor:latest
```

mccheah replied:
Hm, I'm not sure what this extra example gets us that would be helpful for usage that we don't have already - can you elaborate here what the goal is? Having to include spark-defaults.conf as well only makes it seem even more confusing. I mainly modeled the examples off of here: https://spark.apache.org/docs/latest/submitting-applications.html

ash211 reviewed (commented):
Excellent work @mccheah !  I think this is already at a really high level of quality which is awesome.  Docs like these go a long ways towards bringing folks along with the platform.

I also think distributing just the small `Dockerfile` text files themselves and letting the user do with those as they wish sidesteps a lot of issues around Apache publishing, though we'll be discussing that more shortly I think.

Excited to see the progress here!

mccheah committed:
Merge remote-tracking branch 'apache-spark-on-k8s/k8s-support-alternate-incremental' into k8s-docs

mccheah committed:
Address comments, add TODO for things that should be fixed

mccheah commented:
@ash211 @foxish addressed comments. I also scattered TODOs in the docs for things that are documented as the current state of the world but that we want to change in the very near future.

foxish commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,219 @@
+
+## Setting Up Docker Images
+
+In order to run Spark on Kubernetes, a Docker image must be built and available on an accessible Docker registry. Spark
```
There is no direct dependency on Docker, so, s/Docker image/container image/. We can say that in the common case, we need to build docker images and push it to their image registry.

mccheah replied:
Hm, much of the Spark code assumes that Docker is the virtualization framework being used. For example, the image is specified via the configuration setting `spark.kubernetes.driver.docker.image`. Should we also be making all of these configurations more generic?

foxish replied:
I think it's okay for now. I'll discuss this and send a separate PR if we do need to change the wording and the command line args.

foxish commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,219 @@
+
+<!-- TODO master should default to https if no scheme is specified -->
+The Spark master, specified either via passing the `--master` command line argument to `spark-submit` or by setting
+`spark.master` in the application's configuration, must be a URL with the format `k8s://<api_server_url`. Prefixing the
```
k8s://<api_server_url missing closing `>`

foxish commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,219 @@
+The Spark master, specified either via passing the `--master` command line argument to `spark-submit` or by setting
+`spark.master` in the application's configuration, must be a URL with the format `k8s://<api_server_url`. Prefixing the
+master string with `k8s://` will cause the Spark application to launch on a Kubernetes cluster, where the API server is
+contacted  at the appropriate inner URL. The HTTP protocol must also be specified.
```
"at the appropriate inner URL" seems a bit unclear. We can omit that perhaps.

mccheah committed:
Address comments, mostly making images section clearer

mccheah commented:
@foxish made the images section a little more generic and gave special note to Docker as what we provide out of the box. It's very likely that the wording and terminology could be improved there however.

foxish commented:
@mccheah, sg. Thanks! The rest LGTM.

tnachen commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,224 @@
+    bin/spark-submit 
+      --deploy-mode cluster 
+      --class org.apache.spark.examples.SparkPi 
+      --master k8s://https://<k8s-apiserver-host>:<k8s-apiserver-port>
```
Btw is the https necessary?

mccheah replied:
It is for now, later we want to remove it and make https the default, but optionally the user can fill in `http`.

ash211 replied:
Right now these are supported:

- `k8s://https://<host>:<port>`
- `k8s://http://<host>:<port>`

In https://github.com/apache-spark-on-k8s/spark/issues/19 we want to additionally make this supported:

- `k8s://<host>:<port>` which would be equivalent to `k8s://https://<host>:<port>`

ash211 commented:
Looks great to me!  @foxish want to merge it?

foxish commented:
One last nit. Virtual runtime -> Container runtime in the beginning. Will merge once that is fixed.

mccheah committed:
Virtual runtime -> container runtime

mccheah commented:
@foxish done
